### ISSUE TITLE
InvalidCastException when using tuples, yield-return, and switch-expressions together in a particular way

### ISSUE THREAD

**Issue by airbreather** [dotnet/roslyn#45789]
**Version Used**:
.NET Core SDK 3.1.301

**Steps to Reproduce**:

`dotnet build` the following sample application with the listed version of the .NET Core SDK:
# Lib.cs
```csharp
static class Lib
{
    static System.Collections.Generic.IEnumerable<(int, int)> GetVals()
    {
        (int, int) tup = default;
        tup.Item1 |= "" switch { _ => 0 };
        yield return tup;
    }
}
```

# ConsoleApp0.csproj
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>netcoreapp3.1</TargetFramework>
  </PropertyGroup>

</Project>
```

**Expected Behavior**:
It builds a (fairly useless) library.

**Actual Behavior**:
An `InvalidCastException` is thrown.

Full console output: https://gist.github.com/airbreather/657c261b5681171c696bc0e0d853f53c

**Comment by airbreather** [comment]
This can also be seen on SharpLab: https://sharplab.io/#v2:C4LghgzgtgNAJiA1AHwAIEYBsACVAmbAGQEsAjAWACgBvK7e3LRgFgDoMBmAHgApiA7YDGwDgASgB82AOIBTYADUwAGwg8xdBrUoNd2PoOGix2YAFcADtgC82OLIBmYM8uABuTXtOXWASWCyUOjYyLYARGHYEADuxMAAxgAW2NTYAPo2UgAM2AC+HjpejMGoAOzeFgW6uVS5QA==

**Comment by RikkiGibson** [comment]
This crash typically means we've failed to diagnose a ref local which lives across an await/yield boundary. Perhaps related to the way lowering of `|=` works when the LHS is a struct field.

The following code crashes in a similar way:
```cs
struct S1
{
    public int field;
}

static class Lib
{
    static System.Collections.Generic.IEnumerable<S1> GetVals()
    {
        S1 s1 = default;
        s1.field |= "" switch { _ => 0 };
        yield return s1;
    }
}
```

Neither example crashes when `lhs |= rhs` is changed to `lhs = lhs | rhs;` -- probably because we don't try and synthesize a ref to the lhs in the latter case.

The fix may be either to issue a diagnostic or to adjust the lowering of `|=`--we will have to consult the specification.

**Comment by RikkiGibson** [comment]
It appears this was fixed between Roslyn 3.6.0 and 3.7.0
